File: bridge/src/bridge_player.cpp
#if defined(DM_PLATFORM_HTML5)
#include "bridge_player.h"
#include "bridge.h"

int bridge::player::isAuthorizationSupported(lua_State* L) {
    return getBoolean(L, js_bridge_player_isAuthorizationSupported);
}

int bridge::player::isAuthorized(lua_State* L) {
    return getBoolean(L, js_bridge_player_isAuthorized);
}

int bridge::player::id(lua_State* L) {
    return getString(L, js_bridge_player_id);
}

int bridge::player::name(lua_State* L) {
    return getString(L, js_bridge_player_name);
}

int bridge::player::extra(lua_State* L) {
    DM_LUA_STACK_CHECK(L, 1);
    char* extraJson = js_bridge_player_extra();
    dmScript::JsonToLua(L, extraJson, strlen(extraJson));
    free(extraJson);
    return 1;
}

int bridge::player::photos(lua_State* L) {
    DM_LUA_STACK_CHECK(L, 1);
    char* photosJson = js_bridge_player_photos();
    dmScript::JsonToLua(L, photosJson, strlen(photosJson));
    free(photosJson);
    return 1;
}

int bridge::player::authorize(lua_State* L) {
    return makeCallbackWithJson(L, js_bridge_player_authorize, false);
}
#endif

File: bridge/lib/web/bridge_player.js
let js_bridge_player = {
    js_bridge_player_isAuthorizationSupported: function () {
        return bridge.player.isAuthorizationSupported;
    },

    js_bridge_player_isAuthorized: function () {
        return bridge.player.isAuthorized;
    },

    js_bridge_player_id: function () {
        return CStrOrNull(bridge.player.id);
    },

    js_bridge_player_name: function () {
        return CStrOrNull(bridge.player.name);
    },

    js_bridge_player_extra: function () {
        return CStrOrNull(bridge.player.extra);
    },

    js_bridge_player_photos: function () {
        return CStrOrNull(bridge.player.photos);
    },

    js_bridge_player_authorize: function (handler, options, onSuccess, onFailure) {
        var jsOptions = JSON.parse(UTF8ToString(options));
        bridge.player.authorize(jsOptions)
            .then(() => {
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 0, packToJson());
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 1, packToJson(error));
            })
    }
}

mergeInto(LibraryManager.library, js_bridge_player);